chore: output address of uploaded file

jacderida opened the pull request:
- b8487f4f **chore: output address of uploaded file**

  Other minor adjustments for more compact text output.

- b5666cef **feat: provide a `files ls` command**

  Gives the user to list the files they have uploaded, along with their addresses. We were storing
  this data anyway, but it's serialized to binary, so it's not human readable. A little interface on
  the client could be useful for users.

- bdfd4f75 **chore: clarify `files download` usage**

  The documentation for the arguments for the command are clarified, and we introduce a failure if
  only one of the name/address pair is used, as both are necessary.

  Also fix up some code from a complicated rebase.

- ba4809ed **chore: store uploaded files list as text**

  For the uploaded files list, rather than serialize to binary, we use a simple text format. The user
  can now see the addresses of all the files they've uploaded, and use this information to retrieve
  them, if need be. We are also now just using one file, appending to it for each upload.

  Now that this file is human readable, the `files ls` command was deemed unnecessary and was
  therefore removed.

  With respect to the `files` commands, I've also taken the opportunity to make the language in the
  text output more concise and matter of fact. For example, we don't really need to use things like
  exclamation marks in the text.

## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 15 Sep 23 21:01 UTC
This pull request includes the following changes:

- In `cli.rs`:
  - Added a clarification comment for the `timeout` field.
  - Added a clarification comment for the `concurrency` field.
  - Added a clarification comment for the `no_verify` field.

- In `main.rs`:
  - Added a new `FilesCmds` variant, `Ls`, which lists all files uploaded by the current user.
  - Added handling for the `Ls` variant in the `files_cmds` function.

- In `files.rs`:
  - Added a new `FilesCmds` variant, `Ls`, which lists all files uploaded by the current user.
  - Added a new function, `offline_files_cmds`, which handles the `Ls` variant when offline.
  - Added handling for the `Ls` variant in the `files_cmds` function.
  - Added a new function, `list_files`, which lists all files uploaded by the current user.
  - Various changes, including clarifying comments and error handling improvements.

- In `file_apis.rs`:
  - Clarified print statements.

- In `wallet.rs`:
  - Adjusted print statement for completed transfers.
<!-- reviewpad:summarize:end -->

connec commented on sn/src/client/connections/listeners.rs:
```diff
@@ -52,14 +52,12 @@ impl Session {
                     },
-                    Err(Error::Generic(_)) => {
-                        // TODO: FIX error type
+                    Err(_) => {
```
I don't think this changes anything, in the case of `ConnectionLoss`, and makes it more frail otherwise. Currently, when `ConnectionIncoming` yields a `ConnectionLoss` error:

1. It will be returned from `listen_for_incoming_message` (propagated by `?`).
2. It'll hit the `Err(err)` case, and go around the loop again.
3. The next time `incoming_messages.next()` is called, it will return `None`, since `IncomingMessages` will return `None` after connection loss (if not, that's a bug in `qp2p`).
4. On `None`, `listen_for_incoming_message` yields `Error::Generic` and the loop breaks.

The `Error::Generic` stuff does seem a bit convoluted, but we should still filter it to `ConnectionLoss` only imo, or we will throw away connections even if we receive, e.g., a message we can't deserialize.

joshuef replied:
Aye it wont propogate. But we should shut down the loop if it's useless I think. I'll fix that error detection there so e dont trip up on other things though as you say 👍

connec commented on sn/src/client/connections/messaging.rs:
```diff
@@ -298,6 +299,16 @@ impl Session {
         }
 
+        let send_failures = *failed_sends.lock().await;
+        if send_failures >= 2 {
```
I think this could be replaced with `if discarded_responses >= 2 { ... }`. More generally, I think `failed_sends` is redundant, since counting the errors after `join_all` would get the same result with fewer shenanigans?

Also, I thought we could tolerate 2 misbehaving elders, so should this be `> 2`?

connec replied:
Ah, I always forget `task::JoinHandle` yields panics in the outer result. I guess replacing `failed_sends` would need:

```rust
for result in results {
    match result {
        Err(_) => discarded_responses += 1,
        Ok(Err(_)) => failed_sends += 1,
        _ => {}
    }
}
```

e.g., so a bit extraneous for this PR.

joshuef replied:
> if discarded_responses >= 2 { ... }

Hmm, we normally send to 3 elders for queries. We _could_ just send one... but it seemed like perhaps its worth a do over if 2/3 are failing? Dunnnoo

connec replied:
Yeah, that makes sense. I guess this falls under the `TODO` comment above about the approach to validation.

connec commented on sn/src/client/connections/listeners.rs:
```diff
@@ -49,17 +49,28 @@ impl Session {
-                        error!("Error while getting incoming message: {:?}. Listening for next message...", err);
+                    Err(error) => {
+                        match error {
+                            Error::Generic(_) => {
```
Should this not be `Error::IncomingMessages` now, after the change to `listen_for_incoming_message`?

connec commented on sn/src/client/connections/listeners.rs:
```diff
@@ -49,17 +49,28 @@ impl Session {
+                                info!("IncomingMessages listener has closed.");
+                                break;
+                            }
+                            Error::QuicP2p(qp2p_err) => {
```
We could filter for connection loss only with:

```rust
Error::QuicP2p(qp2p::RpcError::Recv(qp2p::RecvError::ConnectionLost(_))) => ...
```

That said, I don't think it would be particularly harmful to close the listener for all `qp2p` errors, just marginally less efficient in some situations (but those would be situations where we or the sender is doing something odd with the streams, like closing them too soon or whatever).

jacderida committed:
chore: output address of uploaded file

Other minor adjustments for more compact text output.

jacderida committed:
feat: provide a `files ls` command

Gives the user to list the files they have uploaded, along with their addresses. We were storing
this data anyway, but it's serialized to binary, so it's not human readable. A little interface on
the client could be useful for users.

jacderida committed:
chore: clarify `files download` usage

The documentation for the arguments for the command are clarified, and we introduce a failure if
only one of the name/address pair is used, as both are necessary.

Also fix up some code from a complicated rebase.

joshuef commented:
I actually think instead of https://github.com/maidsafe/safe_network/commit/050103b85320bac7a68e28395977867514b3985f we should just make that file human readable. 

`files ls` I could see having the implication it may work on any platform, but its just what you have done on this comp... (something that we will likely have later)... Feels like a plaster around just removing the bincode encoding on that file to me.

joshuef reviewed (changes requested):
can we remove https://github.com/maidsafe/safe_network/commit/050103b85320bac7a68e28395977867514b3985f 

I'd rather we made the file human readable and that was how you can check. Doesn't feel like it needs to be in the cli to me.

jacderida commented:
> can we remove [050103b](https://github.com/maidsafe/safe_network/commit/050103b85320bac7a68e28395977867514b3985f)
> 
> I'd rather we made the file human readable and that was how you can check. Doesn't feel like it needs to be in the cli to me.

OK cool. Serialize to JSON?

joshuef commented:
> OK cool. Serialize to JSON?

Just a space separated list perhaps? (newline per file?) 


We could do json, but not sure it's worth it... treat yourself as the consumer of this (you are!) so whatever seems best!

jacderida committed:
chore: store uploaded files list as text

For the uploaded files list, rather than serialize to binary, we use a simple text format. The user
can now see the addresses of all the files they've uploaded, and use this information to retrieve
them, if need be. We are also now just using one file, appending to it for each upload.

Now that this file is human readable, the `files ls` command was deemed unnecessary and was
therefore removed.

With respect to the `files` commands, I've also taken the opportunity to make the language in the
text output more concise and matter of fact. For example, we don't really need to use things like
exclamation marks in the text.
